pkg/certificates/config.go

```diff
@@ -0,0 +1,206 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package certificates
+
+import (
+	"context"
+	"os"
+	"time"
+
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+
+	"storj.io/storj/pkg/identity"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/peertls"
+	"storj.io/storj/pkg/provider"
+	"storj.io/storj/pkg/utils"
+	"storj.io/storj/storage/boltdb"
+	"storj.io/storj/storage/redis"
+)
+
+// CertClientConfig is a config struct for use with a certificate signing service client
+type CertClientConfig struct {
+	AuthToken string `help:"authorization token to use to claim a certificate signing request (only applicable for the alpha network)"`
+	Address   string `help:"address of the certificate signing rpc service"`
+}
+
+// CertServerConfig is a config struct for use with a certificate signing service server
+type CertServerConfig struct {
+	Overwrite          bool   `default:"false" help:"if true, overwrites config AND authorization db is truncated"`
+	AuthorizationDBURL string `default:"bolt://$CONFDIR/authorizations.db" help:"url to the certificate signing authorization database"`
+	MinDifficulty      uint   `default:"16" help:"minimum difficulty of the requester's identity required to claim an authorization"`
+	CA                 identity.FullCAConfig
+}
+
+// SetupIdentity loads or creates a CA and identity and submits a certificate
+// signing request request for the CA; if successful, updated chains are saved.
+func (c CertClientConfig) SetupIdentity(
+	ctx context.Context,
+	caConfig identity.CASetupConfig,
+	identConfig identity.SetupConfig,
+) error {
+	caStatus := caConfig.Status()
+	var (
+		ca    *identity.FullCertificateAuthority
+		ident *identity.FullIdentity
+		err   error
+	)
+	if caStatus == identity.CertKey && !caConfig.Overwrite {
+		ca, err = caConfig.FullConfig().Load()
+		if err != nil {
+			return err
+		}
+	} else if caStatus != identity.NoCertNoKey && !caConfig.Overwrite {
+		return identity.ErrSetup.New("certificate authority file(s) exist: %s", caStatus)
+	} else {
+		t, err := time.ParseDuration(caConfig.Timeout)
+		if err != nil {
+			return errs.Wrap(err)
+		}
+		ctx, cancel := context.WithTimeout(ctx, t)
+		defer cancel()
+
+		ca, err = caConfig.Create(ctx)
+		if err != nil {
+			return err
+		}
+	}
+
+	identStatus := identConfig.Status()
+	if identStatus == identity.CertKey && !identConfig.Overwrite {
```

Comment:
nit - switch can be used here

Reply:
How? The conditions use logical and in and negation. I don't see a way to reduce this to a switch.

Reply:
An example https://github.com/storj/storj/blob/8893884044bcbe636c093e64df2931a5fd6b87b5/pkg/piecestore/psserver/server.go#L270

Reply:
```
switch {
case identStatus == identity.CertKey && !identConfig.Overwrite:
...
case identStatus != identity.NoCertNoKey && !identConfig.Overwrite:
...
default:
...
}